src/serverlist.cpp

```diff
@@ -25,6 +25,71 @@
 #include "serverlist.h"
 
 /* Implementation *************************************************************/
+
+// --- CServerListEntry ---
+CServerListEntry CServerListEntry::parse ( QString strHAddr,
+                                           QString strLHAddr,
+                                           QString sName,
+                                           QString sCity,
+                                           QString strCountry,
+                                           QString strNumClients,
+                                           bool    isPermanent )
+{
+    bool         ok;
+
+    CHostAddress haServerHostAddr;
+    NetworkUtil::ParseNetworkAddress ( strHAddr, haServerHostAddr );
+    if ( CHostAddress() == haServerHostAddr )
+    {
+        // do not proceed without server host address!
+        return CServerListEntry();
+    }
+
+    CHostAddress haServerLocalAddr;
+    NetworkUtil::ParseNetworkAddress ( strLHAddr, haServerLocalAddr );
+    if ( haServerLocalAddr.iPort == 0 )
+    {
+        haServerLocalAddr.iPort = haServerHostAddr.iPort;
+    }
+
+    QLocale::Country lcCountry = QLocale::AnyCountry;
+    int              iCountry  = strCountry.trimmed().toInt ( &ok );
+    if ( ok && iCountry >= 0 && iCountry <= QLocale::LastCountry )
+    {
+        lcCountry = static_cast<QLocale::Country> ( iCountry );
+    }
+
+    int iNumClients = strNumClients.trimmed().toInt ( &ok );
+    if ( !ok )
+    {
+        iNumClients = 10;
+    }
+
+    return CServerListEntry ( haServerHostAddr,
+                              haServerLocalAddr,
+                              CServerCoreInfo ( sName.trimmed().left ( MAX_LEN_SERVER_NAME ),
+                                                lcCountry,
+                                                sCity.trimmed().left ( MAX_LEN_SERVER_CITY ),
+                                                iNumClients,
+                                                isPermanent ) );
+}
+
+QString CServerListEntry::toCSV()
+{
+    QStringList sl;
+
+    sl.append ( this->HostAddr.toString() );
+    sl.append ( this->LHostAddr.toString() );
+    sl.append ( this->strName );
+    sl.append ( this->strCity );
+    sl.append ( QString::number ( this->eCountry ) );
+    sl.append ( QString::number ( this->iMaxNumClients ) );
+    sl.append ( QString::number ( this->bPermanentOnline ) );
+
+    return sl.join ( ";" );
```

Comment:
a server can have a semicolon in its name or city (when it is started from the GUI) and would get sorted out when loading the file

Reply:
Good points.  We should make `;` illegal in the GUI, in that case (and I can strip here), as it's illegal in `--serverinfo`.  That was overlooked when adding the GUI.  Same for newlines (which should never be allowed).

Reply:
A thought, after reviewing #1803 and #1804: those PRs only prevent the server itself creating names containing `\n` and `;`; they don't do anything to prevent older registering servers from sending them to the directory. The code below strips out `\n` but not `;`.

Also, I think there is some code duplicated between those PRs and this one? (e.g. `reIllegal`)

Volker avoided the problem of strange characters in names stored in the ini file by base64 encoding the names and writing in XML format. Maybe there is some code there that could be used here, instead of having to disallow characters?

Reply:
Not a bad idea.  Convert the values to base64, then join with `;`.  It would save having to have the other two PRs, too.

Reply:
Just to note -- the INI file does _not_ use base64 encoding for server name and city:
```
826     // name
827     PutIniSetting ( IniXMLDocument, "server", "name", pServer->GetServerName() );
828
829     // city
830     PutIniSetting ( IniXMLDocument, "server", "city", pServer->GetServerCity() );
```
Again, that's probably an oversight.